e2ee: make fun of the middlebox

fippo opened the pull request:
the visual results of not encrypting the first 10 bytes of the VP8 payload
  https://tools.ietf.org/html/rfc6386#section-9.1
are much more interesting since the decoder gets enough stuff to allow it
to try decoding a bit, resulting in weird images for the middlebox.

fippo committed:
e2ee: make fun of the middlebox

the visual results of not encrypting the first 10 bytes of the VP8 payload
  https://tools.ietf.org/html/rfc6386#section-9.1
are much more interesting since the decoder gets enough stuff to allow it
to try decoding a bit, resulting in weird images for the middlebox.

fippo commented:
@alvestrand for your amusement ;-)

fippo commented:
this sometimes results in pixelation on the receiving end after stopping to decrypt. I suspect what happens there is that a frame which was encrypted and has a checksum is, because the currentCryptoKey is empty, no longer run through the correct parts of the decode function.

Arguably a minor issue but maybe the decode function should look for the "checksum" instead?

alvestrand commented:
I've seen that changing the key sometimes leads to pixelation.
I suspect that what's happening is that there's a frame that's in flight while I change the crypto key, leading to the frame being decrypted with the wrong key.

In a real cryptosystem the solution to this is a) to make sure the checksum is wrong when decrypted with the wrong key, so that it's just frame loss, and b) to synchronize the changing of the key to the frame.

alvestrand commented:
This change should also make the example work without the GFD, right?

fippo committed:
e2ee: deal better with in-flight frames when changing keys

fippo commented:
I don't see how its broken without GFD. Do you have STR? ;-)

Added a second commit to deal a bit better with in-flight frames by dropping them. Still a super simplistic way of dealing with things but the whole scenario of changing keys on the fly is a bit unrealistic anyway. It does avoid pixelation and just causes short freezes which are then resolved by nack/pli (presumably)

alvestrand commented on src/content/peerconnection/endtoend-encryption/js/main.js:
```diff
@@ -114,20 +120,26 @@ function encodeFunction(chunk, controller) {
-    const view = new DataView(chunk.data);
-    const checksum = view.getUint32(chunk.data.byteLength - 4);
-    if (checksum != 0xDEADBEEF) {
+    if (checksum !== 0xDEADBEEF) {
```
To make it more realistic: Have a counter "currentCryptoKeyGeneration" that is incremented every time you call setCryptoKey and add that to the checksum (0xDEADBEEF + currentCryptoKeyGeneration).
That way, frames encrypted with the wrong key will be discarded reliably, and changes from encrypt-with-key-A to encrypt-with-key-B will be handled, instead of just changes from encrypt-with-key-A to no-encrypt and vice versa.

fippo replied:
added as a separate commit. Single byte instead of the four bytes MKI from SRTP ;-)
The console.log shows up nicely when changing the key

alvestrand commented:
actually changing keys on the fly is not only realistic, it's a requirement of group calling :-(

thanks for the update!

If you don't have GFD, and encrypt the whole frame, and set the key before pressing the "call" button, the receiver will see a black frame.

fippo committed:
add keyIdentifier byte to encryption format

reinventing SRTP one field at a time:
  https://tools.ietf.org/html/rfc3711#section-3

alvestrand committed:
Merge branch 'gh-pages' into moar-fun

alvestrand commented:
Merging this. I'm going to add a toggle to encrypt from the start or not, so that we can verify that the GFD matters for keyframe detection.
